tests/macros/tgenast.nim

```diff
@@ -0,0 +1,199 @@
+import std/macros
+from std/strformat import `&`
+import ./mgenast
+
+block:
+  macro bar(x0: static Foo, x1: Foo, x2: Foo, xignored: Foo): untyped =
+    let s0 = "not captured!"
+    let s1 = "not captured!"
+    let xignoredLocal = kfoo4
+
+    # newLit optional:
+    let x3 = newLit kfoo4
+    let x3b = kfoo4
```

Comment:
I am sorry, I have to say no to automatic `newLit` injection. I this this is one of the major problem sources of ``quote do`` and you ported this problem source to ``genAst``. There should be an error message when you try to capture anything that is not of type ``NimNode``.

Reply:
`quote do` doesn't do automatic `newLit` injection, and suffers from bugs like https://github.com/nim-lang/Nim/issues/7375.
I haven't ported this problem to `genAst`, quite the opposite: automatic `newLit` injection as done in this PR not only avoids this category of bugs, but it also simplifies usage, avoiding the redundant `x = newLit x` for every non-NimNode parameter:
`genAst(x1, x2): body`
instead of:
`genAst(x1 = newLit x1, x2 = newLit x2): body` which gets tiring after a while

Note, the `newLit` injection only happens if the type is not `NimNode`, and the user can always provide his own transformation instead, eg:
`genAst(x1 = myNewLit x1): body` and it'll do the right thing

Lastly, there's always `kNoAutoNewLit` in case user wants to avoid automatic `newLit` injection for some reason.

Reply:
I think it is logically wrong to treat `NimNode` and non-`NimNode` differently. Either everything gets newLit applied, but then also NimNode, where expressions are generated to create NimNode values, or nothing. This attitude towards "automatically doing the right thing" usually ends up in "automatically doing the wrong thing". Another question is, what do you do with function symbols? What do you do with type symbols?

Reply:
And no, ``kNoAutoNewLit`` is also yet another flag. Sorry I hate super verbose flags that I have to add to everything. ``genAst`` should not have a flags argument.

Reply:
> genAst should not have a flags argument.

it doesn't, only `genAstOpt` has a flags argument; most code will just use `genAst(x,y): body` and not bother with flags

> And btw kNoAutoNewLit is also yet another crappy flag. Sorry I hate super verbose flags that I have to add to everything

renamed it to `kNoNewLit`, suggestions for alternate names welcome. This flag should very rarely be used so most users will most likely not have to use it


> Either everything gets newLit applied,

that can't work

> or nothing

that'll result in more verbose code littered with `newLit` as discussed above, and the only sane way to do that would be to error when trying to capture anything that is not of type NimNode, however, see point below:

> There should be an error message when you try to capture anything that is not of type NimNode

that would break passing proc symbols, see `bindme6UseExposeFalse` example in mgenast.nim for example:
```nim
macro bindme6UseExposeFalse*(): untyped =
  genAstOpt({kDirtyTemplate}, newStringStream, writeData, readData):
    var tst = "sometext"
    var ss = newStringStream("anothertext")
    writeData(ss, tst[0].addr, 2)
    discard readData(ss, tst[0].addr, 2)
```

> This attitude towards "automatically doing the right thing" usually ends up in "automatically doing the wrong thing"

do you have a realistic concrete example where automatic newLit injection does the wrong thing and an alternative implementation does the correct thing?

> Another question is, what do you do with function symbols? What do you do with type symbols?

function symbols can be captured explicitly, but the capture is optional because of gensym rules (https://nim-lang.github.io/Nim/manual.html#templates-hygiene-in-templates), see `test passing function and type symbols` in tgenast.nim; when `kDirtyTemplate` is used, they must be captured (see `bindme6UseExposeFalse` for an example); it works despite the type not being NimNode, as mentioned above.

type symbols can cause a `Error: internal error: expr(nkType); unknown node kind`, with or without automatic NewLit injection; so auto NewLit injection doesn't make anything worse.

Of course if the function or type symbol is actually of type NimNode (eg a param of the macro, or the result of a `genAst` call (eg
`let Z2 = genAst(): type(true)`), there is nothing special, the capture works as usual

see tgenast.nim `block: #test passing function and type symbols` for an example

Reply:
Without experience, it's hard to know for sure, but the automatic `newLit` application sounds like a reasonable feature to me. You can compare it to `echo` which automatically calls the `$` op for you.

Reply:
First of all, applying newLit to everything is possible. You just need to provide an overload of `newLit` for ``NimNode`` that will generate an ast that will when evaluated, generate the input ast. I did that im my branch. I just didn't call that ``newLit``, because I don't plan to export it (for now).

> do you have a realistic concrete example where automatic newLit injection does the wrong thing and an alternative implementation does the correct thing?

You don't get it. If I merge this PR, I have to make sure that it will work in the future. If I merge this PR it becomes my liability not yours. And want to make sure that there are no bugs introduce because of surprising behavior. So _you_ have to convince me that the automatic newLit injection with the special casing of ``NimNode`` arguments is the most reliable way to do this. And I am not convinced right now.

> type symbols can cause a Error: internal error: expr(nkType); unknown node kind, with or without automatic NewLit injection; so auto NewLit injection doesn't make anything worse.

Internal errors fall into the category of a compiler crash. So they are are bug that need to be addressed.

Reply:
> So they are are bug that need to be addressed.

here's a minimimal test case (independent of this PR), I can create a github issue if needed

```nim
import macros
macro bar(): untyped =
  template baz(T) = echo $T
  result = getAst(baz(int)) # Error: internal error: expr(nkType); unknown node kind
bar()
```

> There should be an error message when you try to capture anything that is not of type NimNode
> First of all, applying newLit to everything is possible

how would that work in the general case? what about closures ?
This works fine despite `fun` not being a `NimNode` (this example works independent of this PR):

```nim
import macros

var count = 0
proc fun() = count.inc
macro bar(): untyped =
  template baz(f) = f()
  result = getAst(baz(fun))
bar()
echo count
```

> And want to make sure that there are no bugs introduce because of surprising behavior

that could be said of any piece of software. I can't prove that this PR is bug free, but I did provide a not-so-small test suite (much more comprehensive than the one for `quote do`) and a fall-back option. If someone runs into a particular scenario where NewLit injection results in error, he can just use `{kNoNewLit}`, and maybe even add a `genAstRaw` template to avoid passing the flag.